src/build/env.py

```diff
@@ -1,312 +1,322 @@
 """
 Creates and manages isolated build environments.
 """
+
 import abc
-import functools
 import logging
 import os
 import platform
 import shutil
-import subprocess
 import sys
 import sysconfig
 import tempfile
 
-from types import TracebackType
-from typing import Callable, Iterable, List, Optional, Tuple, Type
-
-import build
-
+from typing import Dict, Iterable, Sequence, Tuple, Type
 
-if sys.version_info < (3, 8):
-    import importlib_metadata as metadata
-else:
-    from importlib import metadata
-
-try:
-    import virtualenv
-except ModuleNotFoundError:
-    virtualenv = None
+from ._helpers import cache, check_dependency, default_runner
 
 
-_logger = logging.getLogger('build.env')
+_logger = logging.getLogger(__name__)
 
 
 class IsolatedEnv(metaclass=abc.ABCMeta):
-    """Abstract base of isolated build environments, as required by the build project."""
+    """ABC for isolated build environments."""
 
-    @property
     @abc.abstractmethod
-    def executable(self) -> str:
-        """The executable of the isolated build environment."""
+    def __init__(self, isolated_env_manager: 'IsolatedEnvManager') -> None:
+        """
+        :param isolated_env_manager: The ``IsolatedEnvManager`` instance
+        """
         raise NotImplementedError
 
-    @property
     @abc.abstractmethod
-    def scripts_dir(self) -> str:
-        """The scripts directory of the isolated build environment."""
+    def create(self) -> None:
+        """
+        Create the isolated environment.
+
+        This method should be idempotent.
+        """
         raise NotImplementedError
 
     @abc.abstractmethod
-    def install(self, requirements: Iterable[str]) -> None:
+    def install_packages(self, requirements: Iterable[str]) -> None:
         """
-        Install packages from PEP 508 requirements in the isolated build environment.
+        Install packages in the isolated environment.
 
-        :param requirements: PEP 508 requirements
+        :param requirements: PEP 508-style requirements
         """
         raise NotImplementedError
 
+    @abc.abstractmethod
+    def prepare_environ(self) -> Dict[str, str]:
```

Comment:
Possibly allow this to be `None` so that implementers aren't forced to copy the environ if they are not modifying it.